### ISSUE TITLE
Tensor Flow Lite INT8 Issue

### ISSUE THREAD

**Issue by wlimhhn307** [tensorflow/tensorflow#99052]
Hi, I have an issue where i would like to convert the existing .h5 model to float 16 tflite and int8 tflite 
However i faced this issue where for float16 i able to got the result and accuracy drop however for int8 the whole model prediction just overbias and i have checked the overall of my code on the class order. I have tried many approaches and making sure the test data is loaded exactly the same way as how i trained my initial model. Is it because int8 conversion hurt the weightage of the model too much ?
def representative_data_gen():
    train_set.reset()
    samples_per_class = 250 # adjust as needed
    class_counts = {0: 0, 1: 0}
    total_samples = 500
    collected = 0

    while collected < total_samples:
        imgs, labels = next(train_set)
        for img, label in zip(imgs, labels):
            if class_counts[label] < samples_per_class:
                img = tf.expand_dims(img, axis=0)
                yield [tf.cast(img, tf.float32)]
                class_counts[label] += 1
                collected += 1
            if collected >= total_samples:
                break

# Convert to INT8 quantized TFLite model 
converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_data_gen
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.int8
converter.inference_output_type = tf.int8

tflite_quant_int8 = converter.convert()

##And this is the evaluation part 
import numpy as np
from sklearn.metrics import classification_report, confusion_matrix
import tensorflow as tf

# -------- Config --------
tflite_model_path = r"C:\Users\Admin\hicnn\ResNetA_TFLite\Resnet50_quant_int8_test2.tflite"
CLASS_ORDER = ['good', 'bad']   # label 0 = good, 1 = bad
N_PER_CLASS = 5                 # 5 good + 5 bad #I just load few images to test the pipeline

# -------- Load TFLite model --------
interpreter = tf.lite.Interpreter(model_path=tflite_model_path)
interpreter.allocate_tensors()

in_det  = interpreter.get_input_details()[0]
out_det = interpreter.get_output_details()[0]

in_index  = in_det['index']
out_index = out_det['index']
input_scale, input_zero_point = in_det['quantization']
output_scale, output_zero_point = out_det['quantization']

# -------- Collect 5 good + 5 bad from test_set --------
test_set.reset()
good_imgs, bad_imgs = [], []
good_labels, bad_labels = [], []

while (len(good_imgs) < N_PER_CLASS) or (len(bad_imgs) < N_PER_CLASS):
    try:
        imgs, labels = next(test_set)
    except StopIteration:
        break
    for img, lab in zip(imgs, labels):
        if lab == 0 and len(good_imgs) < N_PER_CLASS:
            good_imgs.append(img)
            good_labels.append(0)
        elif lab == 1 and len(bad_imgs) < N_PER_CLASS:
            bad_imgs.append(img)
            bad_labels.append(1)
        if len(good_imgs) >= N_PER_CLASS and len(bad_imgs) >= N_PER_CLASS:
            break

X = np.concatenate([np.stack(good_imgs), np.stack(bad_imgs)], axis=0)
y_true = np.array(good_labels + bad_labels, dtype=int)

# -------- Inference --------
y_pred = []
for i, x in enumerate(X):
    # Quantize input to int8
    x_q = (x / input_scale + input_zero_point).astype(np.int8)
    inp = np.expand_dims(x_q, axis=0)
    interpreter.set_tensor(in_index, inp)
    interpreter.invoke()
    out = interpreter.get_tensor(out_index)
    out = np.squeeze(out)
    prob = float((out - output_zero_point) * output_scale)
    pred = int(prob > 0.5)
    y_pred.append(pred)
    true_name = CLASS_ORDER[y_true[i]]
    pred_name = CLASS_ORDER[pred]
    print(f"Sample {i+1:02d} | True: {true_name.upper():4s} | Pred: {pred_name.upper():4s} | Score(bad)= {prob:.3f}")

# -------- Metrics --------
y_pred = np.array(y_pred, dtype=int)
acc = (y_pred == y_true).mean()
print("\n======================")
print(f"Accuracy on {len(y_true)} images (5+5): {acc:.4f}")
print("======================")
try:
    print("\nClassification Report:")
    print(classification_report(y_true, y_pred, labels=[0,1], target_names=['Good','Bad'], digits=4))
except Exception as e:
    print(f"(classification_report skipped: {e})")
try:
    cm = confusion_matrix(y_true, y_pred, labels=[0,1])
    print("\nConfusion Matrix (rows=true, cols=pred):\n", cm)
except Exception:
    pass

**Comment by MengAiDev** [comment]
**Hey @wlimhhn307,**  

Thanks for sharing the details! INT8 quantization can be tricky—here’s what might help:  

### **Quick Fixes:**  
1. **Try removing these lines** in your converter setup:  
   ```python  
   # Delete these:  
   # converter.inference_input_type = tf.int8  
   # converter.inference_output_type = tf.int8  
   ```  
   **(Forcing INT8 I/O often breaks things—let TFLite handle it automatically.)**  

2. **Clip your input** during inference to avoid overflow:  
   ```python  
   x_q = np.clip(x / input_scale + input_zero_point, -128, 127).astype(np.int8)  
   ```  

3. **Sigmoid the output** (INT8 models return raw logits):  
   ```python  
   prob = 1 / (1 + np.exp(-output_scale * (raw_output - output_zero_point)))  
   ```  

### **Debugging Tips:**  
- Check if all ops quantized properly:  
  ```python  
  print(interpreter.get_tensor_details())  # Look for non-INT8 ops  
  ```  
- Test on **more than 10 samples**—small batches can mislead.  

If it’s still acting weird:  
- Share `model.summary()` (some layers hate INT8).  
- Try `converter.experimental_new_quantizer = True`.  

Let me know how it goes!

**Comment by wlimhhn307** [comment]
okay i will try on this.

**Comment by gaikwadrahul8** [comment]
Hi, @wlimhhn307 
I apologize for the delay in my response, Hi, @MengAiDev Thank you for sharing your pointers here
As far I know please give it try with below option and remove all manual quantization and dequantization logic. The TFLite interpreter will handle it internally if the converter is set up as mentioned below

```
converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_data_gen
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
tflite_quant_int8 = converter.convert()
```
The inference loop should be simplified as follows:
```
y_pred = []
for i, x in enumerate(X):
    # Input is already a float32 array
    inp = np.expand_dims(x, axis=0).astype(np.float32) # Ensure it's float32
    interpreter.set_tensor(in_index, inp)
    interpreter.invoke()
    out = interpreter.get_tensor(out_index)
    # The output will be float32 so no manual dequantization needed.
    # Depending on the model's last layer use argmax for multi-class or a threshold for single output sigmoid.
    # Assuming a binary classification model with two outputs (logits for good/bad)
    pred = np.argmax(out[0])
    y_pred.append(pred)
    true_name = CLASS_ORDER[y_true[i]]
    pred_name = CLASS_ORDER[pred]
    print(f"Sample {i+1:02d} | True: {true_name.upper():4s} | Pred: {pred_name.upper():4s} | Raw output: {out}")
```
Please make sure that data from `representative_data_gen` is pre-processed in the exact same way as the training data including normalization. This is a common oversight. For example, if the Keras model was trained with normalized inputs (0-1), the representative dataset must also be normalized and one more thing larger and more representative dataset allows the converter to accurately determine the min/max ranges for all intermediate activation tensors. If these ranges are off values will be clipped or mapped to incorrect integer values which may be leading to significant information loss and accuracy degradation

Please give it try and let us know is it resolving your issue or not ? If issue still persists please help us with your complete code in Google colab or Github repo along with dataset to reproduce the same behavior from our end. Thank you for your cooperation and understanding.

**Comment by wlimhhn307** [comment]
Hi @MengAiDev @gaikwadrahul8 the issue still same and i try to debug with this code and the output show this where i think the model completely broen once it been convert to int8 tflite. Is there any other way i can fix this? 
 
# === Verify the NEW quantized model works properly ===
print("=== Verifying NEW Quantized Model ===")

# Load the new model
quantint8_file_path = pathlib.Path(r"C:\Users\JAYDEN LIM\hicnn\ResNetA_TFLite\40layers_lr5e-06_bs16.tflite")
interpreter_new = tf.lite.Interpreter(model_path=str(quantint8_file_path))
interpreter_new.allocate_tensors()
inp_details_new = interpreter_new.get_input_details()
out_details_new = interpreter_new.get_output_details()

# Test with the first 3 images from X
print("Testing NEW model with different images:")
for i, img in enumerate(X[:3]):
    inp = np.expand_dims(img, axis=0).astype(np.float32)
    interpreter_new.set_tensor(inp_details_new[0]['index'], inp)
    interpreter_new.invoke()
    tflite_out = interpreter_new.get_tensor(out_details_new[0]['index'])
    prob = float(tflite_out[0][0])
    pred = int(prob > 0.5)
    class_name = class_order[y_true[i]]
    print(f"Image {i+1}: True={class_name.upper()} | Prob={prob:.4f} | Pred={class_order[pred].upper()}")

# Compare with original model on same images
print("\n=== Comparing with Original Model ===")
model = tf.keras.models.load_model(h5_path)
for i, img in enumerate(X[:3]):
    inp = np.expand_dims(img, axis=0)
    orig_prob = model.predict(inp, verbose=0)[0][0]
    pred = int(orig_prob > 0.5)
    class_name = class_order[y_true[i]]
    print(f"Image {i+1}: True={class_name.upper()} | Prob={orig_prob:.4f} | Pred={class_order[pred].upper()}")




=== Verifying NEW Quantized Model ===
Testing NEW model with different images:
Image 1: True=GOOD | Prob=0.9961 | Pred=BAD
Image 1: True=GOOD | Prob=0.9961 | Pred=BAD
Image 2: True=BAD | Prob=0.9961 | Pred=BAD

=== Comparing with Original Model ===
Image 2: True=BAD | Prob=0.9961 | Pred=BAD

=== Comparing with Original Model ===
Image 1: True=GOOD | Prob=0.0262 | Pred=GOOD
Image 2: True=BAD | Prob=0.9776 | Pred=BAD
Image 1: True=GOOD | Prob=0.0262 | Pred=GOOD
Image 2: True=BAD | Prob=0.9776 | Pred=BAD

**Comment by gaikwadrahul8** [comment]
Hi, @wlimhhn307
I believe you've followed suggested steps in my previous comment, May I know which TensorFlow version and Python version  you're using at the moment ? If you're trying latest TensorFlow version then please give it try by downgrading the TensorFlow version and see is it resolving your issue or not ?

Please help us with your complete code in Google colab or Github repo along with dataset to reproduce the same behavior from our end It will be easy to reproduce and investigate this issue further from our end. Thank you for your cooperation and understanding.

**Comment by wlimhhn307** [comment]
Currently i am using this , 
==================================================
SYSTEM INFORMATION
==================================================
Python Version: 3.10.16 | packaged by Anaconda, Inc. | (main, Dec 11 2024, 16:19:12) [MSC v.1929 64 bit (AMD64)]
TensorFlow Version: 2.10.0
NumPy Version: 1.23.5
Pandas Version: 2.2.3

==================================================
GPU INFORMATION
==================================================
GPU Available: [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
Number of GPUs: 1

==================================================
TENSORFLOW BUILD INFO
==================================================
Built with CUDA: True
GPU Device Name: /physical_device:GPU:0
WARNING:tensorflow:From C:\Users\AI-Admin\AppData\Local\Temp\ipykernel_32260\1864361571.py:29: is_gpu_available (from tensorflow.python.framework.test_util) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.config.list_physical_devices('GPU')` instead.
Is GPU Available for TensorFlow: True
==================================================
Is GPU Available for TensorFlow: True
==================================================

**Comment by wlimhhn307** [comment]
@gaikwadrahul8 @MengAiDev I have invited you all to the respitory.

**Comment by gaikwadrahul8** [comment]
Hi, @wlimhhn307 
I apologize for the delayed response, Github repo invitation has expired if possible could you please send me one more time and Just to confirm, is there any specific reason to use older version of TensorFlow ? If not please give it a try with latest stable version of TensorFlow `2.20.0`, You can install it by following this [official documentation](https://www.tensorflow.org/install/pip#linux) as per your OS. Thank you for your cooperation and patience.
